File: datasets/Python/Gemini/Algorithm/medium4.py
def crosswordPuzzle(crossword, words_str):
    """
    Solves a 10x10 crossword puzzle using a backtracking algorithm.
    """
    grid = [list(row) for row in crossword]
    words = words_str.split(';')

    def can_place_horizontal(word, r, c):
        """Checks if a word can be placed horizontally at (r, c)."""
        if c + len(word) > 10:
            return False
        # Check left boundary
        if c > 0 and grid[r][c-1] != '+':
            return False
        # Check right boundary
        if c + len(word) < 10 and grid[r][c+len(word)] != '+':
            return False
        # Check if cells are compatible
        for i in range(len(word)):
            if grid[r][c+i] != '-' and grid[r][c+i] != word[i]:
                return False
        return True

    def can_place_vertical(word, r, c):
        """Checks if a word can be placed vertically at (r, c)."""
        if r + len(word) > 10:
            return False
        # Check top boundary
        if r > 0 and grid[r-1][c] != '+':
            return False
        # Check bottom boundary
        if r + len(word) < 10 and grid[r+len(word)][c] != '+':
            return False
        # Check if cells are compatible
        for i in range(len(word)):
            if grid[r+i][c] != '-' and grid[r+i][c] != word[i]:
                return False
        return True

    def solve(word_index):
        """Recursively places words from the words list."""
        # Base case: All words have been placed
        if word_index == len(words):
            return True

        current_word = words[word_index]
        # Iterate through every cell to find a starting point
        for r in range(10):
            for c in range(10):
                if grid[r][c] == '-' or grid[r][c] == current_word[0]:
                    # Try placing horizontally
                    if can_place_horizontal(current_word, r, c):
                        # Place the word and save the original state of the slot
                        original_slot = [grid[r][c+i] for i in range(len(current_word))]
                        for i in range(len(current_word)):
                            grid[r][c+i] = current_word[i]
                        
                        # Recurse for the next word
                        if solve(word_index + 1):
                            return True
                        
                        # Backtrack: Restore the grid
                        for i in range(len(current_word)):
                            grid[r][c+i] = original_slot[i]

                    # Try placing vertically
                    if can_place_vertical(current_word, r, c):
                        # Place the word and save the original state
                        original_slot = [grid[r+i][c] for i in range(len(current_word))]
                        for i in range(len(current_word)):
                            grid[r+i][c] = current_word[i]

                        # Recurse for the next word
                        if solve(word_index + 1):
                            return True

                        # Backtrack: Restore the grid
                        for i in range(len(current_word)):
                            grid[r+i][c] = original_slot[i]
        
        # If no placement is found for the current word
        return False

    # Start the backtracking process
    solve(0)
    return ["".join(row) for row in grid]

# --- Main execution for manual input ---
if __name__ == '__main__':
    print("Enter the 10x10 crossword grid:")
    crossword_input = []
    for _ in range(10):
        crossword_input.append(input())
    
    print("\nEnter the semicolon-separated words:")
    words_input = input()

    result = crosswordPuzzle(crossword_input, words_input)

    print("\nSolved Crossword:")
    for row in result:
        print(row)

File: datasets/C++/Gemini/Algorithm/medium4.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

// Function to check if a word can be placed horizontally
bool can_place_horizontal(const vector<string>& grid, const string& word, int r, int c) {
    if (c + word.length() > 10) return false;
    if (c > 0 && grid[r][c - 1] != '+') return false;
    if (c + word.length() < 10 && grid[r][c + word.length()] != '+') return false;
    for (size_t i = 0; i < word.length(); ++i) {
        if (grid[r][c + i] != '-' && grid[r][c + i] != word[i]) {
            return false;
        }
    }
    return true;
}

// Function to check if a word can be placed vertically
bool can_place_vertical(const vector<string>& grid, const string& word, int r, int c) {
    if (r + word.length() > 10) return false;
    if (r > 0 && grid[r - 1][c] != '+') return false;
    if (r + word.length() < 10 && grid[r + word.length()][c] != '+') return false;
    for (size_t i = 0; i < word.length(); ++i) {
        if (grid[r + i][c] != '-' && grid[r + i][c] != word[i]) {
            return false;
        }
    }
    return true;
}

// The main recursive backtracking function
bool solve_crossword(vector<string>& grid, const vector<string>& words, int word_index) {
    if (word_index == words.size()) {
        return true; // Base case: all words are placed
    }

    const string& current_word = words[word_index];
    for (int r = 0; r < 10; ++r) {
        for (int c = 0; c < 10; ++c) {
            if (grid[r][c] == '-' || grid[r][c] == current_word[0]) {
                // Try placing horizontally
                if (can_place_horizontal(grid, current_word, r, c)) {
                    vector<string> grid_backup = grid;
                    for (size_t i = 0; i < current_word.length(); ++i) {
                        grid[r][c + i] = current_word[i];
                    }
                    if (solve_crossword(grid, words, word_index + 1)) {
                        return true;
                    }
                    grid = grid_backup; // Backtrack
                }

                // Try placing vertically
                if (can_place_vertical(grid, current_word, r, c)) {
                    vector<string> grid_backup = grid;
                    for (size_t i = 0; i < current_word.length(); ++i) {
                        grid[r + i][c] = current_word[i];
                    }
                    if (solve_crossword(grid, words, word_index + 1)) {
                        return true;
                    }
                    grid = grid_backup; // Backtrack
                }
            }
        }
    }
    return false;
}

int main() {
    cout << "Enter the 10x10 crossword grid:" << endl;
    vector<string> grid(10);
    for (int i = 0; i < 10; ++i) {
        cin >> grid[i];
    }

    cout << "\nEnter the semicolon-separated words:" << endl;
    string words_line;
    cin >> words_line;
    vector<string> words;
    stringstream ss(words_line);
    string word;
    while (getline(ss, word, ';')) {
        words.push_back(word);
    }
    
    solve_crossword(grid, words, 0);

    cout << "\nSolved Crossword:" << endl;
    for (int i = 0; i < 10; ++i) {
        cout << grid[i] << endl;
    }

    return 0;
}